src/lib.rs

```diff
@@ -773,4 +773,258 @@ mod lib_tests {
+        //     ('group',  j, 'seen') where j is less than i
+        for i in 0i64..=13i64 {
+            let mut entry = StateGroupEntry {
+                in_range: true,
+                prev_state_group: prev,
+                state_map: StateMap::new(),
+            };
+            entry
+                .state_map
+                .insert("group", &i.to_string(), "seen".into());
+            entry.state_map.insert("node", "is", i.to_string().into());
+
+            old_map.insert(i, entry);
+
+            prev = Some(i)
+        }
+
+        check_that_maps_match(&old_map, &BTreeMap::new());
+        assert!(true);
+    }
+
+    #[test]
+    fn check_that_maps_match_returns_if_just_old_map_is_empty() {
+        // note that this IS the desired behaviour as only want to ensure that
+        // all groups that existed BEFORE compression, will still collapse to the same
+        // states (i.e. no visible changes to rest of synapse
+
+        let mut new_map: BTreeMap<i64, StateGroupEntry> = BTreeMap::new();
+        let mut prev = None; // note will not be in map
+
+        // This starts with the following structure
+        //
+        // 0-1-2-3-4-5-6-7-8-9-10-11-12-13
+        //
+        // Each group i has state:
+        //     ('node','is',      i)
+        //     ('group',  j, 'seen') where j is less than i
+        for i in 0i64..=13i64 {
+            let mut entry = StateGroupEntry {
+                in_range: true,
+                prev_state_group: prev,
+                state_map: StateMap::new(),
+            };
+            entry
+                .state_map
+                .insert("group", &i.to_string(), "seen".into());
+            entry.state_map.insert("node", "is", i.to_string().into());
+
+            new_map.insert(i, entry);
+
+            prev = Some(i)
+        }
+
+        check_that_maps_match(&BTreeMap::new(), &new_map);
+        assert!(true);
+    }
+
+    #[test]
+    fn check_that_maps_match_returns_if_no_changes() {
+        let mut old_map: BTreeMap<i64, StateGroupEntry> = BTreeMap::new();
+        let mut prev = None; // note will not be in map
+
+        // This starts with the following structure
+        //
+        // 0-1-2-3-4-5-6-7-8-9-10-11-12-13
+        //
+        // Each group i has state:
+        //     ('node','is',      i)
+        //     ('group',  j, 'seen') where j is less than i
+        for i in 0i64..=13i64 {
+            let mut entry = StateGroupEntry {
+                in_range: true,
+                prev_state_group: prev,
+                state_map: StateMap::new(),
+            };
+            entry
+                .state_map
+                .insert("group", &i.to_string(), "seen".into());
+            entry.state_map.insert("node", "is", i.to_string().into());
+
+            old_map.insert(i, entry);
+
+            prev = Some(i)
+        }
+
+        check_that_maps_match(&BTreeMap::new(), &old_map.clone());
+        assert!(true);
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
Actually, what's going on here? `should_panic` tests should really only be used rarely.

Reply:
The should_panic tests should only be testing (hopefully) impossible situations. I added them as they were situations where if the program didn't panic, then the safety checks at the end could pass incorrectly and so database could be corrupted. 

They're panics not results as if any of those situations get reached, then something has gone very wrong with the logic

Reply:
This specific test is just checking that the check_that_maps_match method doesn't just look at predecessors

Reply:
ah ok - I think I could use `#[should_panic(expected = "the panic mesage I want")]` to stop that happening